substrate/frame/revive/src/impl_nonfungibles.rs

```diff
@@ -0,0 +1,293 @@
+
+		let data = IERC721::transferFromCall { from: from_addr, to: to_addr, tokenId: token_id }
+			.abi_encode();
+
+		let ContractResult { result, .. } = Self::bare_call(
+			T::RuntimeOrigin::signed(from),
+			*collection,
+			U256::zero(),
+			GAS_LIMIT,
+			DepositLimit::Balance(<<T as super::pallet::Config>::Currency as fungible::Inspect<
+				_,
+			>>::total_issuance()),
+			data,
+		);
+
+		if let Ok(rv) = result {
+			if rv.did_revert() {
+				Err("Contract reverted".into())
+			} else {
+				Ok(())
+			}
+		} else {
+			Err("Contract out of gas".into())
+		}
+	}
+
+	fn burn(
+		collection: &Self::CollectionId,
+		item: &Self::ItemId,
+		maybe_check_owner: Option<&T::AccountId>,
+	) -> DispatchResult {
+		// Burn simulato: trasferiamo l’NFT al checking account (o a 0x0 se preferisci).
+		let owner = if let Some(acc) = maybe_check_owner {
+			if let Some(current) = <Self as nonfungibles::Inspect<_>>::owner(collection, item) {
+				if &current != acc {
+					return Err("Wrong owner".into());
+				}
+				acc.clone()
+			} else {
+				return Err("Owner not found".into());
+			}
+		} else {
+			<Self as nonfungibles::Inspect<_>>::owner(collection, item).ok_or("Owner not found")?
+		};
+
+		let eth_from = T::AddressMapper::to_address(&owner);
+		let from_addr = Address::from(Into::<[u8; 20]>::into(eth_from));
+
+		let checking = Self::checking_account();
+		let eth_checking = T::AddressMapper::to_address(&checking);
+		let to_addr = Address::from(Into::<[u8; 20]>::into(eth_checking));
+
+		let token_id = EU256::from(*item);
+
+		let data = IERC721::transferFromCall { from: from_addr, to: to_addr, tokenId: token_id }
+			.abi_encode();
+
+		let ContractResult { result, .. } = Self::bare_call(
+			T::RuntimeOrigin::signed(owner),
+			*collection,
+			U256::zero(),
+			GAS_LIMIT,
+			DepositLimit::Balance(<<T as super::pallet::Config>::Currency as fungible::Inspect<
+				_,
+			>>::total_issuance()),
+			data,
+		);
+
+		if let Ok(rv) = result {
+			if rv.did_revert() {
+				Err("Contract reverted".into())
+			} else {
+				Ok(())
+			}
+		} else {
+			Err("Contract out of gas".into())
+		}
+	}
+}
+
+#[cfg(test)]
+mod erc721_contract_tests {
+	use super::*;
+	use crate::{
+		test_utils::{builder::*, ALICE, BOB},
+		tests::{ExtBuilder, RuntimeOrigin, Test},
+		AccountInfoOf, Code,
+	};
+	const ERC721_PVM_CODE: &[u8] = include_bytes!("../fixtures/erc721/erc721.polkavm");
+
+	#[test]
+	fn deploy_erc721_contract() {
+		ExtBuilder::default().existential_deposit(1).build().execute_with(|| {
+			let _ =
+				<<Test as Config>::Currency as fungible::Mutate<_>>::set_balance(&ALICE, 1_000_000);
+
+			let code = ERC721_PVM_CODE.to_vec();
+
+			let Contract { addr, .. } = BareInstantiateBuilder::<Test>::bare_instantiate(
+				RuntimeOrigin::signed(ALICE),
+				Code::Upload(code),
+			)
+			.build_and_unwrap_contract();
+
+			assert!(AccountInfoOf::<Test>::contains_key(&addr));
+		});
+	}
+
+	#[test]
+	fn erc721_balance_of_and_owner_of() {
```

Comment:
The idea of these tests is to see that `<Revive as nonfungibles::Inspect<_>>::owner()` and the rest work